openedx/core/djangoapps/content_libraries/api.py

```diff
@@ -1051,3 +1089,285 @@ def revert_changes(library_key):
         return  # If there is no draft, no action is needed.
     LibraryBundle(library_key, ref.bundle_uuid, draft_name=DRAFT_NAME).cache.clear()
     CONTENT_LIBRARY_UPDATED.send(sender=None, library_key=library_key, update_blocks=True)
+
+
+# Import from Courseware
+# ======================
+
+
+class BaseEdxImportClient(abc.ABC):
+    """
+    Base class for all courseware import clients.
+
+    Import clients are wrappers tailored to implement the steps used in the
+    import APIs and can leverage different backends.  It is not aimed towards
+    being a generic API client for Open edX.
+    """
+
+    EXPORTABLE_BLOCK_TYPES = {
+        "drag-and-drop-v2",
+        "problem",
+        "html",
+        "video",
+    }
+
+    def __init__(self, library_key=None, library=None):
+        """
+        Initialize an import client for a library.
+
+        The method accepts either a library object or a key to a library object.
+        """
+        if bool(library_key) == bool(library):
+            raise ValueError('Provide at least one of `library_key` or '
+                             '`library`, but not both.')
+        if library is None:
+            library = ContentLibrary.objects.get_by_key(library_key)
+        self.library = library
+
+    @abc.abstractmethod
+    def get_block_data(self, block_key):
+        """
+        Get the block's OLX and static files, if any.
+        """
+
+    @abc.abstractmethod
+    def get_export_keys(self, course_key):
+        """
+        Get all exportable block keys of a given course.
+        """
+
+    @abc.abstractmethod
+    def get_block_static_data(self, asset_file):
+        """
+        Get the contents of an asset_file..
+        """
+
+    def import_block(self, modulestore_key):
+        """
+        Import a single modulestore block.
+        """
+
+        # Get or create the block in the library.
+
+        block_data = self.get_block_data(modulestore_key)
+
+        try:
+            library_block = create_library_block(
+                self.library.library_key,
+                modulestore_key.block_type,
+                modulestore_key.block_id,
+            )
+            blockstore_key = library_block.usage_key
+        except LibraryBlockAlreadyExists:
+            blockstore_key = LibraryUsageLocatorV2(
+                lib_key=self.library.library_key,
+                block_type=modulestore_key.block_type,
+                usage_id=modulestore_key.block_id,
+            )
+            get_library_block(blockstore_key)
```

Comment:
>  does it make sense to compose the Blockstore's block usage_id with the course_key something like: usage_id=hash(modulestore_key.course_key + modulestore_key.block_id)?

That's a solution that I've used before for similar problems, and I think it would work well, but it does add a layer of complexity/indirection that can make it harder to debug issues with content (because you can't easily tell from the resulting `block_id` which course it came from).

I would suggest that we either:
(1) Put some explicit comments in the code explaining that this is meant to be idempotent, and when the command is run, count the number of existing blocks and just print a message like ("17 blocks already existed in the destination library and were not imported."). We don't necessarily have to worry about or support the situation where users import from multiple courses to the same library, as long as we're clear about what we're doing and the potential issues for that use case.
or
(2) Do what you're suggesting but use `usage_id = modulestore_key.block_id + "_" + hash(modulestore_key.course_key)[0,3]` so that humans can look at the ID and tell what the original course-scoped block ID was.

Reply:
I like your suggested approach. Deal.